chore(release): 3.17.0

sfoster1 opened the pull request:
This is release 3.17.0.

shlokamin committed:
feat(protocol-designer): reset time fields in pause step (#4950)

fix #4949

mcous committed:
style(protocol-designer): change default exports to named exports (#4939)

Co-authored-by: Ian Solano-Kamaiko <2371418+iansolano@users.noreply.github.com>

sfoster1 committed:
refactor(api): split module labware handling (#4931)

This puts all the module labware stuff (module geometries, loading, etc)
into its own file module_geometry.py. This is both a good idea in
general since it reduces the size of the fairly monstrous labware.py and
also paves the way for future work in loading modules.

Closes #4955

Commit:
feat(protocol-designer): allow modules to be placed on compatble labware (#4972)

* feat(protocol-designer): allow modules to be placed on compatble labware

closes #4726

* clean up test file

* fix typo

Kadee80 committed:
fix(protocol-designer): Update conditional experimental settings warning text (#4978)

closes #4911

Commit:
feat(api): Define existing endpoints using FastAPI/Pydantic (#4962)

* initial checkin of FastAPI port.
currently uses uvicorn to run: uvicorn opentrons.app.main:app.

* completed the networking endpoint definitions

* completed the control endpoints

* settings endpoints added

* add deck calibration endpoints

* add useFastApi feature flag to conditionally start fastapi server from app directory instead of aiohttp app.

* fix lint errors

* set fastapi, uvicorn, and python-multipart versions

* add restart_required=True to useFastApi advanced setting.

* use HTTPStatus enum for status codes

* fastapi uses hardware object created by application bootstrap process

* improve working of useFastApi feature flag description. make it very discouraging.

closes #4917

sanni-t committed:
feat(api): add wifi disconnect capability (#4957)

and add disconnect endpoint to openapi.json

mcous committed:
refactor(app): simplify top-level Epic type (#4951)

b-cooper committed:
refactor(api): hw controller dir for clarity and avoiding circ deps (#4986)

shlokamin committed:
fix(protocol-designer): set magnetic module params for both pristine … (#4990)

* fix(protocol-designer): set magnetic module params for both pristine and unprestine forms

fix #4908

* Add default null module id for magnet steps

Kadee80 committed:
fix(protocol-designer): Update SlotMap and CrashInfoBox styles (#4996)

closes #4989

Commit:
fix(protocol-designer): round caption under engage height to match input box (#4993)

* fix(protocol-designer): round caption under engage height to match input box

closes #4912

* fix null heights improperly displaying

shlokamin committed:
fix(protocol-designer): removes error when fn cannot find temperature or thermocycler on deck (#4994)

fix #4841

shlokamin committed:
fix(protocol-designer): fix failing action test (#5002)

sfoster1 committed:
fix(api): fix recent change breaking api server init (#5003)

A recent commit broke server initialization by passing
initialize_robot() into asyncio.loop.run_until_complete, but it's a
synchronous function.

IanLondon committed:
fix(shared-data): fix typo in v4 protocol schema (#4997)

Commit:
test(protocol-designer): rename test-with-flow folder to __tests__  (#5018)

* chore(protocol-designer): rename test-with-flow folder to __tests__

closes #4983

* update jest config to ignore some folders inside test folders

* move some folders around to follow other app structures

* update configs for fixtures and utils folder

IanLondon committed:
fix(protocol-designer): null out deleted module IDs from step forms (#5017)

* also fix form-level errors for missing moduleId in temperature/magnet/pause step forms

Closes #4883

mcous committed:
refactor(app): add actions + state for GET /wifi/list to networking module (#5023)

Closes #5021

sfoster1 committed:
refactor(shared-data): deck: add v2 schema (#4961)

This adds a new deck definition schema, data for it, and tests. This
schema allows specification of module types for per-slot compatibility.

Adding tests also exposed that the v1 deck fixture did not conform to
the v1 schema; this is also fixed.

Closes #4956

sfoster1 committed:
chore(merge): master->edge for 3.16.0

Conflicts fixup: apply the
api/src/opentrons/hardware_control/__init__.py changes from
c42d4dc62938ccaad9135d0ed153e5ba35c99381 to
api/src/opentrons/hardware_control/api.py after
7d3ea9822c297614761a65beb3499454b599837f moved most of the code there.

mcous committed:
refactor(discovery-client): switch from default to named exports (#4992)

sfoster1 committed:
refactor(shared-data): split v2 module defs into own files (#4936)

This split makes the schema apply to only an individual module
definition, which will in turn allow definition consumers to schema
check definition inputs on an individual basis instead of as a document
defining all valid modules.

SyntaxColoring committed:
fix(app): Deck cal: Tell users to attach an Opentrons tip, not a GEB tip. (#5048)

Kadee80 committed:
chore(*): Separate feature request and bug report issue templates (#5041)

mcous committed:
refactor(app): switch app from default exports to named exports (#4975)

mcous committed:
build(js): upgrade babel and webpack toolchains (#4959)

Commit:
feat(protocol-designer): warn user when exporting a file with an unused pipette and module (#5046)

* feat(protocol-designer): add warnings when user exports a file with an unused pipette or module

closes #4563

* add more tests

* fix flow typings

* make inexact types exact object types

* suggested change to merge the 2 getUnusedEntities fns

* clean up test file

Co-authored-by: Ian London <ianlondondu@gmail.com>

mcous committed:
refactor(components): Use named exports and exact prop types (#5027)

mcous committed:
refactor(js): remove remaining default exports and add error to lint config (#5057)

Closes #4830

IanLondon committed:
feat(protocol-designer): disambiguate module model vs type (#5029)

closes #4897

IanLondon committed:
feat(protocol-designer): unify "button row" styles (#5056)

Addresses #1670. Doesn't close it b/c I didn't want to mess with the lost-grid button rows now

Not a pure refactor: this does slightly change styles (margin above buttons)

IanLondon committed:
fix(protocol-designer): fix tests for FileSidebar after merging #5029 (#5077)

sfoster1 committed:
chore(merge): master->edge for 3.16.1

mcous committed:
refactor(app): add wifi configure actions and epics to networking module (#5074)

mcous committed:
refactor(js): Upgade Flow to v0.119.1 and fix type errors (#5058)

Co-authored-by: Ian London <ianlondondu@gmail.com>

mcous committed:
refactor(app): add GET /wifi/keys to app/src/networking (#5087)

b-cooper committed:
feat(api): invert management of hw control thread (#5078)

Remove thread management from sync adapter's responsibilities and manage singular thread for all hw control concerns in a new class. Also, nuke SingletonAdapter.

Closes #4870

SyntaxColoring committed:
fix(api): Avoid Python 3.8 unsupported `str in Enum` expression (#5088)

Kadee80 committed:
feat(protocol-designer): Add V4 export warning modal (#5092)

closes #4934

ewagoner committed:
refactor(protocol-designer): cypress updates (#5101)

* build(cypress): update cypress.io to 4.0.2

* test(protocool-designer): fix broken cypress test due to dom changes

Commit:
feat(scripts): python_build_utils supports upcoming api split  (#5113)

* to support the process of splitting up api into two projects add opentrons to python_build_utils.

* allow running python_build_utils from the scripts directory

* api remains as the opentrons library name. the api webservice is renamed to robot-server.

* fix wrong prefix for robot server

Commit:
fix(api): ThreadManager hides builder exceptions (#5108)

* ThreadManager constructor can return happily even if the build function raised an exception. We add a check in the constructor for a valid managed object once the thread initialization event is signaled. An exception is raised.

* fix lint issue

Commit:
chore(api): Upgrade api to python 3.7 (#4999)

* Upgrade api to python 3.7

* update api docs to indicate we are going from py3.6.4 to py3.7.6

* change contributing to indicate 3.7.6 of python

closes #4991

Commit:
fix(api): add ThreadManager.cleanup() calls in test fixtures (#5109)

mcous committed:
refactor(app): add POST /wifi/keys and GET /wifi/eap-options to `networking` (#5093)

btmorr committed:
docs(api): update example apiVersion to 2.2 (#5054)

Commit:
refactor(api): create a method that will initialize api.(#5124)

a slight refactor of main to create a method that will initialize the api. the purpose is to provide a method for the robot_server to initialize and use the opentrons api.  

in the future, api.main and api.server.main will be removed. all http server business will happen in robot_server.

IanLondon committed:
chore(js): add jest lint rules and fix tests (#5114)

fix problems with tests that these rules revealed

Commit:
style(protocol-designer): adjust font styling on slot warning text (#5105)

* style(protocol-designer): adjust font styling on slot warning text

closes #5075

* use variable for line height

Commit:
feat(protocol-designer): create modal component to show announcements (#5091)

* feat(protocol-designer): create modal component to show announcements for new versions

closes #4995

* add tests

* update styling of modal to match design shape

* make version match pd version instead of run app

* update modal to match new styling

* move announcement key to persist file with other keys

IanLondon committed:
feat(protocol-designer): save v3/v4 files correctly (#5100)

Closes #4919

IanLondon committed:
fix(protocol-designer): fix typo in deck setup explanation modal (#5104)

Closes #5102

b-cooper committed:
chore(app, protocol-designer): update to latest redux version (#5125)

update redux dependency to v4.0.5 for app and protocol designer

Closes #5066

iansolano committed:
build(opentrons): add styled-components to buildtools (#5129)

* build(opentrons): add styled-components to buildtools

mcous committed:
fix(app): accept mixed-case extensions for protocols and custom labware (#5153)

Closes #5151

mcous committed:
refactor(app,protocol-designer): replace date format library moment with date-dns (#5127)

Closes #4269

Kadee80 committed:
refactor(protocol-designer): Update pause item substep messaging (#5149)

closes #5116

mcous committed:
fix(app): build RPC pipette tip rack list from both containers and instruments (#5147)

This works around a bug from the RPC API in PAPIv1 where a pipette may be reported
to use no tip racks even if a tip rack itself reports usage by that same pipette

b-cooper committed:
feat(api): create a centralized hw execution manager (#5138)

Make a single manager to be shared across children of a hardware controller that handles the flow and interruption of execution at runtime.

Closes #4871

chazmanc committed:
docs(labware-designer): correct x/y docs param names (#5042)

* labware designer readme update

* prettier

Co-authored-by: Laura Cox <31892318+Laura-Danielle@users.noreply.github.com>

Commit:
feat(api): robot-server sub project created  (#5130)

* copy api to api-server. first step in creating a separate api-server.

* move webservice packages to api from src/opentrons

* add fastapi dependencies

* fastapi app imports

* add editable api as dev dependency.
add aiohttp dependency

* fix imports

* server tests pass

* refactor api startup

* move deck calibration end points to api.aiohttp.endpoints.deck_calibration from opentrons.deck_calibration.endpoints

* rename api-server to robot-server

* update Config.in to robot-server

* rename root package from api to robot_server.
update dependencies.
let mypy find the editable installation of api lib.

* use robot-server in python build utils call

* only report coverage in source directory

* move tests/opentrons/server to tests/robot_server/server

* fix serving openapi spec

* add integration/test_server

* delete non webservice related source and tests

* src path in makefile

* add robot-server to mono repo makefile

* fix failing tests

* add atomicwrites windows dependency

* read me is more webservice focused

* remove docs and makefile dircetives related to docs

* add on_shutdown handler to stop the MainRouter threads

* Remove global session from deck_calibration. Create a SessionWrapper that allows access to the single SessionManager instance. This allows cleaning up session when creating a new one.

* remove jsonrpcserver dependency. pylama is a dev dependency.

* add package.json

* refactor main to use aiohttp app from opentrons package instead of robot-server

* remove aiohttp from robot-server. we will import the aiohttp from opentrons lib

* rename fastapi package to service

* remove openapi validation in robot-server. it's all in the api package.

* minimal unit tests for robot_server

* clean up comments

* add robot-server to eslintignore

* use opentrons.main.initialize to create hardware

* update setup.py to build robotserver

* update manifest.in for robot server

* remove markers from pytest.ini

* buildroot changes for robot-server.
service file modified to launch correct entry.

* updates to Makefile:
remove pypi related directives.
remove simulate.

* remove resources and shared data from Makefile

* fixes to robot-server buildroot build

* remove open api from robot-server MANIFEST

* fix make push

* type annotation corrections. version fixes. and other minor stuff.

* remove pypi related api docs. other minor cleanup.

* address some PR comments.

* address PR comments

* Bug fix. Don't use OTAPI_DUMP_BR_VERSION in mk file. Also don't use ot_api_name.

closes #5011 #5012

ahiuchingau committed:
fix(api): remove unnecessary pipette movement during mix (#5128)

* fix(api): remove unnecessary pipette movement during mix

closes #4640

b-cooper committed:
fix(api): compare smoothie version before hardware controller on boot (#5165)

Before creating the hardware controller that the robot server will use, let's now fetch the current fw version directly from the driver in order to check for a smoothie update. This prevents a race condition that leads to zombie module instances. Modules that were created in the first hardware controller instance were not fully cleaned up by the time the new modules were instantiated and grabbed the locked drivers. Then the first modules get garbage collected and those locked drivers become disconnected.

Commit:
feat(protocol-designer): add border to module tag when hovering over step (#5161)

* feat(protocol-designer): add border to module tag when hovering over a step in timeline

closes #5110

* change styling for tag due to border

* remove border from left/right depending on tag location

* fix failing test

* fix css issue with border

* fix foreignObject logging warning in tests

* remove unused classname

b-cooper committed:
fix(api): make api coroutine properties sync for polls during pause (#5166)

The hardware api had a couple of coroutine properties that wanted to be awaited inside of endpoints that are constantly polled during the run of a protocol.  With a central hardware thread, a pause will lead to the smoothie driver blocking the thread and sometimes causing the loop to time out. For now we can prevent this edge condition from crashing the session by ensuring polled endpoints only access sync properties and don't call into the hw event loop.

Co-authored-by: Seth Foster <seth@opentrons.com>

IanLondon committed:
refactor(js): convert jest "test" to "it" (#5115)

sfoster1 committed:
feat(api): support gen2 modules (#5039)

Gen2 modules (with revisions above 20) are now detected as such. We use
the V2 module defs (if the protocol API version is high enough) to load
the modules, respecting the compatibility elements in the definitions.
We respect and apply the transformations.

This also adds the new moduleModel key to GET /modules and the model
element to the RPC module model.

Closes #4960

Commit:
feat(api): robot server replaces api (#5139)

* remove service file.
modify readme to indicate this is a library not the api server.
move api server stuff out of Makefile.

* remove fastapi and related dependencies

* port SessionWrapper from robot_server's deck_calibration back to api

* remove fastapi from api package. it is now in robot-server.

* api doesn't install a service anymore

* add robot-server to mono repo's buildroot config.

* deck calibration: Due to ThreadManager change, SessionWrapper should not call join on old instance

* bring back make push

* remove fastapi app launch. it's been moved to robot-server.

* fix make dev in api. echos a deprecation warning and runs make dev from robot-server

closes #5049

Commit:
refactor(robot-server): udpate control model (#5181)

Commit:
refactor(api): port aiohttp health unit test to fastapi (#5163)

* adding requests as a dev dependency. required by starlette's TestClient

* add conftest.py with basic api_client fixture that overrides the get_hardware dependency with a hardware simulator.

* port aiohttp health endpoint unittests to fastapi

Commit:
fix(robot-server): fix failure in fw_version reading in health endpoint. (#5185)

shlokamin committed:
feat(api): add start set temperature to api (#5179)

* feat(api): add start set temperature to api

#5176

mcous committed:
feat(app): report PAPI version during protocol analytics events (#5053)

Closes #4971

Kadee80 committed:
fix(protocol-designer): Update temperature modules disengage text (#5159)

closes #5111

Commit:
feat(robot_server): add missing endpoints to robot_server from api (#5167)

* add missing endpoints to fastapi: update firmware, wifi disconnect, and get logs.

* missing doc strings.

Commit:
test(protocol-designer): fix failing cypress tests following addition of announcement modal (#5183)

IanLondon committed:
feat(js): update Formik to v2 (#5190)

* refactor Formik forms to conform to v2 changes
* add flow-typed defs for Formik 2.x.x, remove Formik type stub

Commit:
fix(robot-server): cannot import log control (#5207)

b-cooper committed:
fix(api): make hardware control gantry fns observe execution manager (#5180)

In order to prevent pause/cancel edge case failures, make hardware control move and home commands observe the execution manager. Also, augment the execution manager to provide conditional locked access to a state variable, that now has a canceled state in addition to running and paused.

sfoster1 committed:
fix(api): call module functions across threads (#5194)

Moving to the threadmanager broke the case of calling module functions
through the threadmanager, since the module objects would be returned
directly.

They are now wrapped in CallBridger objects that have the
__getattribute__ override the same as thread managers but don't do the
actual thread managing. This is LRU cached so we don't build a bunch of
them.

sfoster1 committed:
feat(app, components): Support gen2 modules (#5177)

This PR
- introduces support for gen2 modules by supporting the new structures
of the HTTP and RPC endpoints from the robot (while handling backwards
compatibility)
- Changes most of the app's structure to support the new version 2
schema for module and deck definitions

Closes #4960

Co-authored-by: IanLondon <ianlondondu@gmail.com>

sfoster1 committed:
feat(api): updated images for all modules (#5210)

These now have different images for v2 modules, as well as updated
images for the originals that match what shipped in terms of visible
stickers and style.

Closes #4224

ahiuchingau committed:
feat(api): use instrument max achievable height in plan_moves (#5193)

* feat(api): use instrument max achievable height in plan_moves

closes #5156

Commit:
feat(protocol-designer): allow custom labware on modules (#5175)

* feat(protocol-designer): allow custom labware on modules on deck

closes #4910

* add tests

* clean up files and tests

* add more tests

shlokamin committed:
feat: include all vscode configs in gitignore (#5222)

Commit:
refactor(robot-server,api): robot server port networking fast api (#5184)

* create system/wifi module and move some functions from server/endpoints/networking there.

* more refactoring

* code clean up and restructuring.

* move eap option validation from network endpoints to wifi.
rewrote test_key_lifecycle to test the key api. not the http api.

* reworked test_networking_endpoints to only test the http interaction with the key functions. Not the underlying file system. That belongs in the test_wifi module.

* test_network_status added to nmcli. this didn't belong in the server endpoint tests as it is really testing our ability to communicate with nmcli.

* remove deprecated comment

* start robot_server networking endpoints

* progress

* keys is a reserved word. use an alias.

* list keys tests pass

* networking tests pass minus the disconnect

* disconnect tests

* tests for wificonfiguration model validation

* add todo note regarding wifi disconnect

* fix lint errors

* cannot import log_control.MAX_RECORDS due to syslog import failure.

* clean up the instances of multiple response types.

* rename V1ErrorResponse to V1BasicResponse. It isn't always used for errors. Name was misleading.

* fix bad test

* fix lint error

* fix bugs that were revealed when running on PI.
Configure didn't work because we weren't testing how the nmcli.configure call was made.
get wifi/keys failed due to validation errors on pydantic model due to attribute named keys. also, someone (me) didn't port the aiohttp unit tests properly.

closes #5143

Kadee80 committed:
feat(protocol-designer): Add module prefix to labware (#5205)

closes #5119

IanLondon committed:
feat(protocol-designer): add modal to add a pause after set temp (#5182)

Closes #5117

* refactor add/select step selectors for mocking, and add tests

sfoster1 committed:
chore: add risk assessment field to pr template (#5233)

In a risk assessment, you go over your change set and identify other
areas of the system it might touch. Particularly important in a broad
and connected system, this process both gets the person submitting the
PR to consider the impact of their changes and points testers and
reviewers in the right direction.

Commit:
refactor(robot-server, api): implement module endpoints in fastapi (#5229)

* add hardware fixture

* use mock hardware for health test

* port module unit tests from aiohttp to fastapi

* move module entpoints and modules out of control and into their own python files.

* minor fixes in api

* create find_matching_module function and use it.

* use V1BasicResponse

closes #5141

Commit:
refactor(robot-server, api): add /pipettes endpoint and tests (#5237)

* add /pipettes endpoint and tests
use V1BasicResponse

* rename Pipette model to the more accurate PipettesByMount

closes #5145

Commit:
feat(protocol-designer): remove warnings and display correct version for gen 2 modules (#5224)

* feat(protocol-designer): remove warnings and display correct version of module image on file page

closes #5152

* remove crash warning from edit modules and add tests

* remove comment

* make crashable warning text only show for the crashable module

* fix flow error

IanLondon committed:
feat(protocol-designer): allow user to bypass typeform modal (#5169)

merge hotfix changes from PD release 3.0.8

Closes #5126

mcous committed:
build(styled-components): add styled-components to app deps (#5236)

sfoster1 committed:
fix(api): properly parse escaped nmcli responses (#5246)

nmcli uses ':' as a field separator. If it encounters a : in a field
value, it escapes it with a backslash. We weren't parsing this
correctly (or indeed at all) and therefore couldn't connect to wifi
networks whose names include a colon. Parsing escaped strings is hard,
so this change uses the csv module to parse them properly

mcous committed:
refactor(app): add client-side state for Wi-Fi disconnect (#5234)

iansolano committed:
feat(robot-server): add formatted json responses and error handling (#5208)

* feat(robot-server): add formatted json responses and error handling

SyntaxColoring committed:
feat(labware-library): Support more pipettes in custom labware test protocol (#5084), and use opentrons_96_tiprack_20ul instead of 10ul

sfoster1 committed:
fix(app): disable module commands when protocol paused (#5209)

The module command cards displayed in the protocol run page have
historically allowed users to send module commands (like setting
temperature) when the protocol was active but paused. That used to work
essentially as a side effect: modules didn't respect pausing, so you
could tell them to do things while the protocol was paused.

Now that modules respect pausing, sending module commands when the
protocol is paused results in the modules not doing anything.

The science and UX intent of controlling modules from the protocol run
page is to implement pre-protocol preparation (like preheat) or
post-protocol behavior (like keeping some incubation going), not to
control the modules when paused - and in fact, when more advanced module
control behaviors land in protocol designer or in the python api,
controlling modules when the protocol is paused could break protocol
execution.

This PR changes the module command cards to only work when the protocol
is not paused or running.

When the module cards are disabled, the action button is now displayed but inaccessible, with an explanatory tooltip on hover.

Co-authored-by: Mike Cousins <mike@opentrons.com>

Commit:
refactor(robot-server): add /motors endpoints and tests (#5242)

* add /motors endpoints and tests

* create motor models from opentrons.hardware_control.types.Axis

* fix transform_validation_error_to_json_api_errors. error['loc'] may contain non strings.

closes #5192

IanLondon committed:
fix(protocol-designer): fix bug with selecting magnet > disengage step (#5257)

Closes #5247

sfoster1 committed:
refactor(api): fullstep during plunger unstick (#5245)

It turns out that bumping the current isn't enough to unstick pipettes
that have been on the shelf for a really long time. We also need to
configure the smoothie to drive the pipette axes at full steps.

To do this, we introduce some new gcodes and wrap motion commands in
them.

This also restricts the move split / unsticks to pipette axes, since
only the pipette axes have configurable step ratios.

Also, bump split current to 1.75A

Closes #5168

sfoster1 committed:
refactor(api): increase gen2 multi flow rate (#5266)

These bump the flow rates from 5mm/s plunger speed to 10 mm/s plunger
speed per hw suggestion.

Closes #5264

IanLondon committed:
feat(protocol-designer): clear out engageHeight (#5254)

Clear out engageHeight from MAGNET steps when magnetic module model is
edited.

Closes #5225

IanLondon committed:
docs(shared-data): update v4 protocol schema - engageHeight description (#5259)

Closes #5227

IanLondon committed:
feat(shared-data): correct existing labware defs engageHeight (#5261)

Closes #5226

Commit:
refactor(robot-server, api): robot server port /robot endpoints fast api (#5252)

* robot endpoint progress: home, light, and position implemented with unittests.

* add test for mount with z too low to aiohttp

* move api implemented and unit tested. moved camera into its own router.

* lint issues fixed

* move ThreadAsyncLock into hardware_control package.

* rename to threaded_async_lock and at it to hardware_control export

* use ThreadedAsyncLock in home and move handlers.

* fix lint errors

* add model validator tests. more clean up.

closes #5191

sfoster1 committed:
fix(api): make sure we move plunger to bottom after home (#5253)

This commit make a new function _do_plunger_home which homes a plunger
on a mount, if it exists, and then moves that plunger to its configured
bottom. This is now used both by API.home_plunger (which was where this
behavior was before) but also by a call to API.home() that involves
pipettes, ensuring that plungers are always moved to the bottom after
they are homed.

Closes #5164

sfoster1 committed:
fix(api): return tips to tipracks from the same height as drop tip (#5187)

Avoids an issue where the ejector would contact the tip rack body.

Closes #5186

IanLondon committed:
feat(protocol-designer): use different ranges for gen1 vs gen2 engageHeight (#5262)

Closes #5230

* introduce `.meta` to `HydratedFormData`

mcous committed:
fix(api): fix server crash due to unresponsive motor controller at boot (#5277)

sfoster1 committed:
fix(api): reset location cache before tip probe (#5269)

When tip probes fail, subsequent retries would see the robot lowering
the pipette dangerously to move back to the front of the robot for a new
tip. This was a combination of two issues:
- The location cache associated with the protocol context needed to be
cleared
- The smoothie driver automatically homes after errors like tip probe
failures, but doesn't tell the hardware controller, so it needs to
forcibly request the new position after a failure.

Closes #4793

IanLondon committed:
feat(protocol-designer): allow setting initial settings w env vars (#5268)

Allow E2E tests etc the option to set initial values of feature flags (aka experimental settings) via env vars.

Laura-Danielle committed:
feat(api): Add API calibration check session (#5197)

sfoster1 committed:
refactor(build): add robot-server to top level push target (#5282)

You can now do `make push` in the root of the repo and it will push the
robot server as well as the api and update server. previously it would
only do api and update, and you'd have to do a separate push to get the
robot server on there.

iansolano committed:
fix(robot-server): fix typing_extensions bug introduced in json response work (#5281)

fix typing_extensions bug by removing Literal type and subsequently the library

closes #5275

Commit:
feat(protocol-designer): allow selection of gen 2 modules and use formik (#5265)

* feat(protocol-designer): allow selection of gen 2 modules and use formik for form handling of page

closes #5150

* use model options from constants

Commit:
refactor(api,robot-server): Robot server port camera fast api (#5278)

* move ffmpeg picture taking into new camera module under system.

* implement take picture endpoint with tests

* StreamingResponse accepts a background param which will be called when sending is done. This is the correct way to clean up captured image file. Not the daft way I was doing it.

* take_picture uses Path instead of str for filename.
remove redundant checking of paths existing.

* make sure file is closed before deleting. thanks again Windows.

closes #5223

shlokamin committed:
feat(api): add v4 JSON executor (#5221)

* Add support for V4 JSON schema in v4 executor
* Move temperature module await temperature command down to module level
* Update versioning notes
* Updated API docs to mention only two pipettes are supported
* Add start_set_temp and await_temp to run log

Commit:
chore(protocol-designer): update crash warning images (#5267)

* chore(protocol-designer): update crash warning images

closes #5112

* update images

sfoster1 committed:
fix(api): exec protocol contents in module mode (#5292)

if you give exec() separate locals and globals, it executes the provided
data as if embedded in a class definition which we very much do not want
- we want it to be executed as if in just a file, which of course it is.
This makes scoping work the same as it does when people write their
protocols and expect them to work.

While we're at it, call the argument we feed to the protocol's run
method something a little less likely to collide.

Closes #4982
Closes #4981

Laura-Danielle committed:
refactor(api): Add in a GET request for calibration check session (#5291)

ahiuchingau committed:
refactor(api): limit gen2 mag module max engage height (#5244)

* refactor(api): limit gen2 mag module max engage height

closes #5212

ahiuchingau committed:
refactor(api): use different labware engage heights for GEN1 and GEN2 modules (#5271)

closes #5214, closes #5215

sfoster1 committed:
chore(docs): document needing to home in advanced running (#5293)

Closes #4973

mcous committed:
refactor(app): Rewrite Wi-Fi connection flows to use new HAPI state (#5280)

Closes #4849

Co-authored-by: Sanniti <sanniti.p@nyu.edu>
Co-authored-by: Ian Solano-Kamaiko <2371418+iansolano@users.noreply.github.com>

mcous committed:
feat(app): add Wi-Fi disconnect to app (#5296)

Laura-Danielle committed:
refactor(api): Add success message to delete endpoint (#5299)

sfoster1 committed:
feat(app): enable gen2 multi pipettes (#5297)

Closes #4698

sfoster1 committed:
chore(release): 3.17.0-alpha.0 (#5300)

mcous commented on app-shell/build/release-notes.md:
```diff
@@ -1,10 +1,24 @@
+  Opentrons App is now disabled when the protocol is paused.
+  
+## New Features
+- OT-2s that are on at least version 3.17.0 and connected via USB can now be
```
No Wi-Fi actions (including disconnect) are currently gated behind being connected over the wired connection. It's definitely our recommendation, but the UI does not currently warn nor stop you if you're only connected via Wi-Fi

See #2551

sfoster1 replied:
OK. Let's leave it in for now I guess

sfoster1 committed:
fix(api): fix blocking module tasks (#5315)

Closes #5312

ahiuchingau committed:
fix(api): fix update module firmware (#5317)

closes #5314

b-cooper committed:
fix(app): alignment and time formatting in thermocycler card (#5320)

mcous committed:
fix(app): fix tiprack in tip probe instructions when tipracks are shared (#5319)

Closes #5311

sfoster1 committed:
chore(bump): 3.17.0-alpha.1 (#5323)

ahiuchingau committed:
fix(api): correctly handle magdeck engage kwargs (#5330)

sfoster1 committed:
fix(api): move splitting: strip step postfix (#5337)

the motor controller board ignores commands that start with a space (?)
and so it wasn't restoring the old steps/mm setting and flipping back
the microstepping switch after moves.

ahiuchingau committed:
fix(api): fix missing arg in v1 move splitting during cache instr models (#5338)

sfoster1 committed:
chore(release): 3.17.0-alpha.2 (#5342)

sfoster1 committed:
chore(release): 3.17.0-beta.0 (#5357)

nusrat813 reviewed (approved):
👍

Kadee80 reviewed (approved):
🏭
